src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -1125,6 +1125,113 @@ .maxstack  2
   IL_0013:  ret
   IL_0014:  ldc.i4.0
   IL_0015:  ret
+}");
+        }
+
+        [Fact]
+        [WorkItem(24550, "https://github.com/dotnet/roslyn/issues/24550")]
+        [WorkItem(1284, "https://github.com/dotnet/csharplang/issues/1284")]
+        public void ConstantPatternVsUnconstrainedTypeParameter01()
+        {
+            var source =
+@"using System;
+class Program
+{
+    static void Main()
+    {
+        Console.WriteLine(Test1<object>(null));
+        Console.WriteLine(Test1<int>(1));
+        Console.WriteLine(Test1<int?>(null));
+        Console.WriteLine(Test1<int?>(1));
+
+        Console.WriteLine(Test2<object>(0));
+        Console.WriteLine(Test2<int>(1));
+        Console.WriteLine(Test2<int?>(0));
+        Console.WriteLine(Test2<string>(""frog""));
+
+        Console.WriteLine(Test3<object>(""frog""));
+        Console.WriteLine(Test3<int>(1));
+        Console.WriteLine(Test3<string>(""frog""));
+        Console.WriteLine(Test3<int?>(1));
+    }
+
+    public static bool Test1<T>(T t)
+    {
+        return t is null;
+    }
+    public static bool Test2<T>(T t)
+    {
+        return t is 0;
+    }
+    public static bool Test3<T>(T t)
+    {
+        return t is ""frog"";
+    }
+}";
+            var compilation = CreateCompilation(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyDiagnostics();
+            var expectedOutput =
+@"True
+False
+True
+False
+True
+False
+True
+False
+True
+False
+True
+False
+";
+            var compVerifier = CompileAndVerify(compilation, expectedOutput: expectedOutput);
+            compVerifier.VerifyIL("Program.Test1<T>(T)",
+@"{
+  // Code size       10 (0xa)
+  .maxstack  2
+  IL_0000:  ldarg.0
+  IL_0001:  box        ""T""
```

Comment:
Why do we need a box? It looks like this is generated because we choose object as the type of the pattern expression. Is there a way to avoid boxing?

Reply:
As far as I know, no. The CLR does not actually produce code for the box. If it is a reference type, the box goes away. If it is a value type, the whole method turns into `false`.

---
In reply to: [180594436](https://github.com/dotnet/roslyn/pull/25995#discussion_r180594436) [](ancestors = 180594436)